src/libraries/System.Private.CoreLib/src/System/Single.cs

```diff
@@ -2167,7 +2167,7 @@ public static bool TryParse(ReadOnlySpan<byte> utf8Text, NumberStyles style, IFo
         static ushort IBinaryFloatParseAndFormatInfo<float>.NormalMantissaBits => SignificandLength;
         static ushort IBinaryFloatParseAndFormatInfo<float>.DenormalMantissaBits => TrailingSignificandLength;
 
-        static int IBinaryFloatParseAndFormatInfo<float>.MinFastFloatDecimalExponent => -65;
+        static int IBinaryFloatParseAndFormatInfo<float>.MinFastFloatDecimalExponent => -45;
```

Comment:
These aren't the same as `MinDecimalExponent`, which would be `-45` for float and `-324` for double.

Instead, it's a separate value taken from Daniel Lemire's `FastFloat` paper which is explained here-ish: https://github.com/fastfloat/fast_float/blob/ff7cb4426320d8297b2e9e01d8fd2723d7368469/include/fast_float/fast_table.h#L19-L26

```cpp
/**
* The smallest non-zero float (binary64) is 2^-1074.
 * We take as input numbers of the form w x 10^q where w < 2^64.
 * We have that w * 10^-343  <  2^(64-344) 5^-343 < 2^-1076.
 * However, we have that
 * (2^64-1) * 10^-342 =  (2^64-1) * 2^-342 * 5^-342 > 2^-1074.
 * Thus it is possible for a number of the form w * 10^-342 where
 * w is a 64-bit value to be a non-zero floating-point number.
 *********
 * Any number of form w * 10^309 where w>= 1 is going to be
 * infinite in binary64 so we never need to worry about powers
 * of 5 greater than 308.
 */
 ```

Reply:
Yes I noticed that project, but did not find this description of this value.

Reply:
I'm a bit confused around this. It seems that `MinFastFloatDecimalExponent` for Half should be Floor(Log10(-14-10-64))=-27, but it's specified as -8 instead. Values less than the exponent will still be caught by slow path. Is this value intentional for Half?

Reply:
It's entirely possible that `Half` was pessimized a bit and whoever added the support also misunderstood or misremembered what this value was meant to be.

I don't have the actual logic for the math here paged into my head at the moment, but at a glance what you have sounds correct.

---------

Trying to page it in...

The `MinDecimalExponent` is the negative of `floor(log10(2^-(MaxExponent + SignificandLength))`, thus:
* `Half:     floor(log10(2^-(15   + 11))) = floor(log10(2^-26))   = -8`
* `BFloat16: floor(log10(2^-(127  + 8)))  = floor(log10(2^-135))  = -41`
* `Single:   floor(log10(2^-(127  + 24))) = floor(log10(2^-151))  = -46`
* `Double:   floor(log10(2^-(1023 + 53))) = floor(log10(2^-1076)) = -324`

Since epsilon is `2^-(MaxExponent + TrailingSignificandLength - 1)` and since we round midpoints to even, this means the exactly representable value must be greater than halfway between `2^-1075` and `2^-1074` for double since the former becomes zero itself.

The `MinFastDecimalExponent` is then `-floor(x)` given `((2^64 - 1) * 10^-x) < 2^-(MaxExponent + TrailingSignificandLength)`:
  * `x > ((MaxExponent + TrailingSignificandLength) * log(2) + log(3) + log(5) + log(17) + log(257) + log(641) + log(65537) + log(6700417)) / (log(2) + log(5))`
  * `x > ((MaxExponent + TrailingSignificandLength) * log(2) + 44.361419) / 2.302585`

Thus, given the above:
* `Half:     (x > (25   * log(2) + 44.361419) / 2.302585) = (x > 26.7917)`
* `BFloat16: (x > (134  * log(2) + 44.361419) / 2.302585) = (x > 59.6039)`
* `Single:   (x > (150  * log(2) + 44.361419) / 2.302585) = (x > 64.4204)`
* `Double:   (x > (1075 * log(2) + 44.361419) / 2.302585) = (x > 342.873)`

So we get:
* `Half:     -floor(26.7917) = -26`
* `BFloat16: -floor(59.6039) = -59`
* `Single:   -floor(64.4204) = -64`
* `Double:   -floor(342.873) = -342`

This is then used [here](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs,1280) where we have `if ((w == 0) || (q < TFloat.MinFastFloatDecimalExponent)) { return default; }` given `w` is the `decimal mantissa` and `q` is the `decimal exponent`, which flows since the values given represent the inclusive minimum decimal exponent that can still produce a non-zero result. -- This notably means `Half` is potentially incorrect for some input and `Single` is more inclusive than necessary by `1` today.

Reply:
Thanks, this also matches my calculation. I hope comment about getting these values will help people.